test/test_torch.py

```diff
@@ -12340,6 +12340,172 @@ def test_against_np(tensor, bins=100, min=0, max=0):
+            else:
+                self.assertEqual(actual[1], torch.from_numpy(expected[1]).to(torch.get_default_dtype()))
+
+        def test_against_np_custombins(tensor, bins, weights=None, density=False, rtol=None, atol=None):
+            nparr = tensor.cpu().numpy()
+            npbins = bins.cpu().numpy()
+            if weights is not None:
+                npweights = weights.cpu().numpy()
+            else:
+                npweights = None
+            actual = torch.histogram(tensor, bins=bins, weights=weights, density=density)
+            expected = np.histogram(nparr, bins=npbins, weights=npweights, density=density)
+            self.assertEqual(actual[0], torch.from_numpy(expected[0]), rtol=rtol, atol=atol)
+            self.assertEqual(actual[1], torch.from_numpy(expected[1]))
+
+        if TEST_NUMPY:
+            default_dtype = torch.get_default_dtype()
+
+            for test_default_dtype in [torch.float, torch.double]:
+                torch.set_default_dtype(test_default_dtype)
+                test_against_np_uniformbins(torch.tensor([1., 2, 1], device=device))
+                test_against_np_uniformbins(torch.randn(5000, device=device))
+
+                # test integer input
+                test_against_np_uniformbins(torch.tensor([1, 2, 3, 4], device=device))
+
+                # test weights arg
+                test_against_np_uniformbins(torch.randn(5000, device=device), weights=torch.rand(5000, device=device))
+                test_against_np_uniformbins(torch.randn(5000, device=device), weights=torch.randint(100, (5000, ), device=device))
+
+                # test density arg
+                test_against_np_uniformbins(torch.randn(5000, dtype=torch.double, device=device), density=True)
+
+                # test weights and density arg
+                test_against_np_uniformbins(torch.randn(5000, dtype=torch.double, device=device),
+                                            weights=torch.rand(5000, dtype=torch.double, device=device), density=True)
+                test_against_np_uniformbins(torch.randn(5000, dtype=torch.double, device=device),
+                                            weights=torch.randint(100, (5000, ), dtype=torch.double, device=device), density=True)
+
+                # Test bins arg
+                test_against_np_uniformbins(torch.randn(301, device=device), bins=20)
+
+                # Test truncated range
+                test_against_np_uniformbins(torch.randn(701, device=device), min=0.1, max=1)
+
+                noncontig = torch.randn(100, 3, device=device)[:, 2]
+                test_against_np_uniformbins(noncontig)
+
+                multidim = torch.randn(3, 5, 7, 2, device=device)
+                test_against_np_uniformbins(multidim)
+
+                expanded = torch.randn(1, 5, 1, 2, device=device).expand(3, 5, 7, 2)
+                test_against_np_uniformbins(expanded)
+
+                test_against_np_custombins(torch.tensor([1., 2, 1], device=device),
```

Comment:
See suggestions for uniform bins above. I think it'll make this test much easier to read and more maintainable.